web/web/models/alert.py

```diff
@@ -1,28 +1,235 @@
+    
+    alert_type_id = Column(db.Integer, 
+                           db.ForeignKey('alert_types.alert_type_id'), 
+                           nullable=False)
+    
+    assigned_to = Column(db.Enum(AssignedToOptions),  
+                         nullable=False)
+
+    description = Column(db.Text,
+                         nullable=False)
+
+    status = Column(db.Enum(Status),
+                    nullable=False)
+
+    context = Column(db.Enum(ContextType),
+                     nullable=False)
+
+    context_id = Column(db.String(64),
+                        nullable=False)
+
+    resolution = Column(db.Text,
+                        nullable=False)
+    
+    updated_at = Column(TIMESTAMP, 
+                        nullable=False,
+                        server_default=text('CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP'))
+
+    created_at = Column(TIMESTAMP,
+                        nullable=False,
+                        server_default=func.now())
+
+    # Methods
+    def __repr__(self):
+        return f'<Alert alert_id={self.alert_id} alert_type_id={self.alert_type_id} created_at={self.created_at}>'
+
+    def get_filter_type(self):
+        '''Returns filter type as a string for alert frontend set-up'''
+
+        if self.alert_type.name.value == 'Import Capacity' \
+            or self.alert_type.name.value == 'Export Capacity' \
+                or self.alert_type.name.value == 'Load Yellow' \
+                    or self.alert_type.name.value == 'Load Red':
+            return 'Capacity bounds'
+        
+        if self.alert_type.name.value == 'Price Yellow' \
+            or self.alert_type.name.value == 'Price Red' \
+                or self.alert_type.name.value == 'Price':
+            return 'Price alerts'
+        
+        if self.alert_type.name.value == 'Telecomm':
+            return 'Telecomm alerts'
+        
+        if self.alert_type.name.value == 'Resource':
+            return 'Resource depletion'
+        
+        if self.alert_type.name.value == 'Peak Event':
+            return 'Peak event'
+
+# Relationship declared on other table (dependency on table import for event listener)
+AlertType.alerts = relationship('Alert',
+                                 backref=db.backref('alert_type'))
+
+# Event listener for email notifications
+@event.listens_for(Alert, 'after_insert')
+def after_insert(mapper, connection, target):
+    '''Sends email after a new alert event is inserted into database'''
+
+    # Query all notifications that are active for the alert type
+    notifications = Notification.query \
+                                .filter(Notification.alert_type_id==target.alert_type_id, Notification.is_active==True) \
+                                .all()
+
+    # Checks if there are no notifications for early exit
+    if notifications == []:
+        return None
+
+    # Since after_insert event is triggered before relationships are configured, alert_type relationship needs to be declared
```

Comment:
I'm not sure what this comment means. If you have the id, then the relationship exists right?

Reply:
after_insert gets triggered before relationships are mapped. There are a variety of trigger events for sqlalchemy (I tried using some others that would be after relationships are mapped but did not function as expected like after_insert did. I can look back again at the other trigger event options.

Reply:
Hummmm, that seems like a bizarre side-effect of this event listener. Did you look into Flask's signals as an alternative?